File: 0496-next-greater-element-i/0496-next-greater-element-i.cpp
class Solution {
public:
    vector<int> nextGreaterElement(vector<int>& nums1, vector<int>& nums2) {
        unordered_map<int, int> mp;
        vector<int> ans(nums1.size(),-1);

        int m = nums1.size();
        int n = nums2.size();

        for (int i = 0; i < nums2.size(); i++) {
            mp[nums2[i]] = i;
        }

        for (int i = 0; i < m; i++) {
            int idx = mp[nums1[i]];
            for (int j = idx + 1; j < n; j++) {
                if (nums2[j] > nums1[i]) {
                    ans[i]=nums2[j];
                    break;
                }
            }
        }

        return ans;
    }
};

File: 0496-next-greater-element-i/0496-next-greater-element-i.java
class Solution {
    public int[] nextGreaterElement(int[] nums1, int[] nums2) {
        int m = nums1.length;
        int n = nums2.length;

        int[] ans = new int[m];

        HashMap<Integer, Integer> mp = new HashMap<>();

        Stack<Integer> st = new Stack<>();

        for (int i = 0; i < n; i++) {
            while (!st.empty() && st.peek() < nums2[i]) {
                mp.put(st.peek(), nums2[i]);
                st.pop();
            }

            st.push(nums2[i]);
        }

        for (int i = 0; i < m; i++) {
            if (mp.containsKey(nums1[i]))
                ans[i]=mp.get(nums1[i]);

            else
                ans[i]=-1;
        }

        return ans;
    }
}